File: trestle-reasoner/src/main/java/com/nickrobison/trestle/types/TrestleFact.java
package com.nickrobison.trestle.types;

import com.nickrobison.trestle.types.temporal.TemporalObject;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.io.Serializable;

/**
 * Created by nrobison on 10/16/16.
 */
public class TrestleFact<T> implements Serializable {
    private static final long serialVersionUID = 42L;

    private final String identifier;
    private final String name;
    private final T value;
    private final TemporalObject validTemporal;
    private final TemporalObject databaseTemporal;
    private final @Nullable String language;
    private final Class<?> javaClass;

    @SuppressWarnings({"dereference.of.nullable"})
    public TrestleFact(String identifier, String name, T value, @Nullable String language, TemporalObject validTemporal, TemporalObject databaseTemporal) {
        this.identifier = identifier;
        this.name = name;
        this.value = value;
        this.validTemporal = validTemporal;
        this.databaseTemporal = databaseTemporal;
        this.language = language;
        this.javaClass = value.getClass();
    }

    public String getIdentifier() {
        return identifier;
    }

    public String getName() {
        return name;
    }

    public T getValue() {
        return value;
    }

    /**
     * Returns the language of the Fact, if its a multi-language string
     *
     * @return - Optional of Fact language
     */
    public @Nullable String getLanguage() {
        return this.language;
    }

    public Class<?> getJavaClass() {
        return this.javaClass;
    }

    public TemporalObject getValidTemporal() {
        return validTemporal;
    }

    public TemporalObject getDatabaseTemporal() {
        return databaseTemporal;
    }

    @Override
    public String toString() {
        return "TrestleFact{" +
                "identifier='" + identifier + '\'' +
                ", name='" + name + '\'' +
                ", value=" + value +
                ", validTemporal=" + validTemporal +
                ", databaseTemporal=" + databaseTemporal +
                ", language='" + language + '\'' +
                ", javaClass=" + javaClass +
                '}';
    }
}

File: trestle-server/src/main/webapp/app/ExploreModule/visualize/individual/trestle-fact.ts
import { ITrestleTemporal, TrestleTemporal } from "./trestle-temporal";
import { IInterfacable } from "../../interfacable";

export interface ITrestleFact {
    identifier: string;
    name: string;
    type: string;
    value: string;
    databaseTemporal: ITrestleTemporal;
    validTemporal: ITrestleTemporal;
}

export class TrestleFact implements IInterfacable<ITrestleFact> {
    private identifier: string;
    private name: string;
    private type: string;
    private value: string;
    private databaseTemporal: TrestleTemporal;
    private validTemporal: TrestleTemporal;

    constructor(fact: ITrestleFact) {
        console.debug("Building fact", fact.identifier);
        this.identifier = fact.identifier;
        this.name = fact.name;
        this.type = fact.type;
        this.value = fact.value;
        this.databaseTemporal = new TrestleTemporal(fact.databaseTemporal);
        this.validTemporal = new TrestleTemporal(fact.validTemporal);
    }

    public getID(): string {
        return this.identifier;
    }

    public getName(): string {
        return this.name;
    }

    public getValue(): string {
        return this.value;
    }

    public getType(): string {
        return this.type;
    }

    public getValidTemporal(): TrestleTemporal {
        return this.validTemporal;
    }

    public getDatabaseTemporal(): TrestleTemporal {
        return this.databaseTemporal;
    }

    public isSpatial(): boolean {
        return this.name === "asWKT";
    }

    public asInterface(): ITrestleFact {
        return {
            identifier: this.identifier,
            name: this.name,
            type: this.type,
            value: this.value,
            databaseTemporal: this.databaseTemporal.asInterface(),
            validTemporal: this.validTemporal.asInterface()
        };
    }
}
